### PR TITLE
test(retrieval): unit-test hybrid_search temporal post-filter, rerank_depth widening, and rrf_merge edge cases

### LINKED ISSUE #325: [super-qa] high: Bi-temporal post-filter in `hybrid_search` never exercised at unit level
> ⚠️ Generated by `/super-qa` 2026-06-01. Severity & details from agent analysis; **not independently source-verified**. Review before action.

**Bi-temporal post-filter in `hybrid_search` never exercised at unit level**

- **ID:** `search/testing-hybrid-temporal-postfilter`
- **Severity:** high
- **Area:** retrieval · **Type:** test
- **Location:** `src/search/hybrid.rs:218-230`
- **Auto-fixable:** no

## Description
The `valid_at`/`t_valid`/`t_invalid` post-filter block at `hybrid.rs:218-230` is the only part of `hybrid_search` that enforces bi-temporal semantics. Every unit test in `hybrid.rs` passes `valid_at: None` and constructs facts with `t_valid: None, t_invalid: None`. The three branches — (1) future-dated `t_valid > cutoff` filtering out a fact, (2) past-invalidated `t_invalid <= cutoff` filtering out a fact, (3) explicit `valid_at` time-travel revealing a future fact — are all absent from `hybrid.rs` inline tests. Coverage exists only indirectly via `tests/eval/conformance/temporal_consistency.rs` going through the full `execute_query` stack, not through `hybrid_search` directly.

## Suggested approach
Add three tests to `hybrid.rs::tests`:
```rust
#[test]
fn hybrid_search_excludes_future_t_valid() {
    // Insert fact with t_valid = now + 1 hour; verify it is absent from results with valid_at=None
}
#[test]
fn hybrid_search_excludes_past_t_invalid() {
    // Insert fact with t_invalid = now - 1 hour; verify it is absent
}
#[test]
fn hybrid_search_valid_at_time_travel() {
    // Insert fact with t_valid = future; query with valid_at = future + 1s; verify it is included
}
```

## References
—

### LINKED ISSUE #478: [super-qa] medium: `rrf_merge` missing edge-case tests: both-empty, k=0, duplicate IDs within one list
> ⚠️ Generated by `/super-qa` 2026-06-01. Severity & details from agent analysis; **not independently source-verified**. Review before action.

**`rrf_merge` missing edge-case tests: both-empty, k=0, duplicate IDs within one list**

- **ID:** `search/testing-rrf-edge-cases`
- **Severity:** medium
- **Area:** retrieval · **Type:** test
- **Location:** `src/search/hybrid.rs:111-126`
- **Auto-fixable:** no

## Description
Three edge cases for `rrf_merge` (a pure, public function) are untested:
1. **Both lists empty** — no test for `rrf_merge(&[], &[], k)`, which should return `vec![]`.
2. **k=0** — proptest range at line 508/527/540 is `1..120u32`, so k=0 is never exercised. At k=0 the denominator is `0 + rank + 1`, a valid result, but unverified.
3. **Duplicate IDs within a single list** — e.g. `[(1, -1.0), (1, -2.0)]` in the FTS list; `scores.entry(id)` accumulates the score twice, which may be an unintended boost. No test documents the intended behavior.

## Suggested approach
Add three example-based tests:
```rust
#[test]
fn rrf_merge_both_empty() {
    assert_eq!(rrf_merge(&[], &[], 60), vec![]);
}
#[test]
fn rrf_merge_k_zero() {
    let merged = rrf_merge(&[(1, -1.0)], &[], 0);
    assert_eq!(merged.len(), 1);
    assert!((merged[0].1 - 1.0).abs() < 1e-10); // 1/(0+0+1) = 1.0
}
#[test]
fn rrf_merge_duplicate_ids_within_fts() {
    // Document: duplicate id contributes its score twice
    let merged = rrf_merge(&[(1, -1.0), (1, -2.0)], &[], 60);
    assert_eq!(merged.len(), 1); // deduped by HashMap
}
```
Also extend the proptest `k` range to `0..120u32`.

## References
—

### LINKED ISSUE #480: [super-qa] medium: `rerank_depth` widening logic untested at `hybrid_search` unit level
> ⚠️ Generated by `/super-qa` 2026-06-01. Severity & details from agent analysis; **not independently source-verified**. Review before action.

**`rerank_depth` widening logic untested at `hybrid_search` unit level**

- **ID:** `search/testing-rerank-depth-unit`
- **Severity:** medium
- **Area:** retrieval · **Type:** test
- **Location:** `src/search/hybrid.rs:143-145`
- **Auto-fixable:** no

## Description
The invariant at line 144 — `effective_target = rerank_depth.unwrap_or(limit).max(limit)` (can only widen, never shrink) — is exercised by one test in `src/engine/tests.rs:1952` that goes through the full engine stack. The four unit tests inside `hybrid.rs` all pass `rerank_depth: None`. Two specific behaviors have no unit coverage in `hybrid.rs`: (1) `rerank_depth < limit` is clamped to `limit` and (2) `rerank_depth > limit` widens the candidate pool above the limit before truncation.

## Suggested approach
Add two unit tests inside `hybrid.rs::tests`:
```rust
#[test]
fn rerank_depth_smaller_than_limit_clamped_to_limit() {
    // rerank_depth=1, limit=3; results should still be clamped to 3 (not 1)
}
#[test]
fn rerank_depth_larger_than_limit_widens_pool() {
    // rerank_depth=9, limit=3; result count should be <= 3 but candidates
    // evaluated before truncation should be drawn from rerank_depth window
}
```

## References
—

### PR DESCRIPTION
**Pull request by dutiona**
Part of #257.

Fills unit-level coverage gaps in `search/hybrid.rs` that were previously only exercised through the full engine/conformance stack (or not at all).

## #325 — bi-temporal post-filter in `assemble_results`
Adds a `make_fact_temporal` fixture override (the existing helpers hardcode `t_valid: None, t_invalid: None`, so the `Utc::now()` cutoff never filtered) and three tests driving the temporal branches directly:
- future-dated `t_valid > cutoff` exclusion,
- past `t_invalid <= cutoff` exclusion,
- explicit `valid_at` time-travel making a future-dated fact visible.

## #480 — `rerank_depth` widening invariant
Two `hybrid_search`-level tests for `effective_limit = rerank_depth.unwrap_or(limit).max(limit)`:
- `rerank_depth < limit` clamps back up to `limit` (does not shrink),
- `rerank_depth > limit` widens the candidate pool above `limit` before truncation.

## #478 — `rrf_merge` edge cases
- both-empty inputs → `vec![]`,
- `k=0` → denominator `0+rank+1`, rank-0 scores `1.0`,
- duplicate id within one list → documents the accumulator double-count (summed contributions, single HashMap output row),
- extend the proptest `k` range from `1..120` to `0..120`.

## Gate
- `cargo clippy -p memory-engine --all-targets --all-features -- -D warnings` — clean
- `cargo test -p memory-engine` — 1259 passed (+8 new); `search::hybrid` 14 → 22, also green under `--features ann`

Closes #325, Closes #480, Closes #478.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: memory-engine/lib/memory-engine/src/search/hybrid.rs [hunk 1]
```diff
@@ -877,6 +940,160 @@ mod tests {
...
+    // supplies the explicit bounds these three branches need.
+
+    #[test]
+    fn hybrid_search_excludes_future_t_valid() {
+        // A fact whose valid-time starts in the future (`t_valid > cutoff`) is a
+        // scheduled/not-yet-true fact: it must NOT surface in a regular query
+        // (valid_at = None → cutoff = now).
+        let conn = setup();
+        let store = FactStore::new(&conn, DIM);
+        let future = Utc::now() + chrono::Duration::hours(1);
```
[the remark this answers is not in the document]
Comment (dutiona) [reply memory-engine/lib/memory-engine/src/search/hybrid.rs:958]: `chrono::Duration` is the established majority convention in this codebase (69 uses vs 40 for `TimeDelta` across `src/`) and is a re-exported type alias for `chrono::TimeDelta` — behaviour is identical. The new tests use `Duration` for consistency with that majority and with the already-merged #835 expired-probe tests (which also use `chrono::Duration`). Keeping it as-is; no change. Resolving.

File: memory-engine/lib/memory-engine/src/search/hybrid.rs [hunk 2]
```diff
@@ -877,6 +940,160 @@ mod tests {
...
+        assert_eq!(results[0].fact.content, "present rust fact");
+    }
+
+    #[test]
+    fn hybrid_search_excludes_past_t_invalid() {
+        // A fact invalidated in the past (`t_invalid <= cutoff`) is no longer
+        // true at query time and must be excluded.
+        let conn = setup();
+        let store = FactStore::new(&conn, DIM);
+        let past = Utc::now() - chrono::Duration::hours(1);
```
[the remark this answers is not in the document]
Comment (dutiona) [reply memory-engine/lib/memory-engine/src/search/hybrid.rs:985]: `chrono::Duration` is the established majority convention in this codebase (69 uses vs 40 for `TimeDelta` across `src/`) and is a re-exported type alias for `chrono::TimeDelta` — behaviour is identical. The new tests use `Duration` for consistency with that majority and with the already-merged #835 expired-probe tests (which also use `chrono::Duration`). Keeping it as-is; no change. Resolving.

File: memory-engine/lib/memory-engine/src/search/hybrid.rs [hunk 3]
```diff
@@ -877,6 +940,160 @@ mod tests {
...
+
+    #[test]
+    fn hybrid_search_valid_at_time_travel() {
+        // Time-travel: an explicit `valid_at` in the future moves the cutoff so a
+        // future-dated fact (`t_valid <= valid_at`) becomes visible, while the
+        // same fact stays invisible to a now-cutoff query (proved by the
+        // `excludes_future_t_valid` test above).
+        let conn = setup();
+        let store = FactStore::new(&conn, DIM);
+        let future_valid = Utc::now() + chrono::Duration::hours(1);
```
[the remark this answers is not in the document]
Comment (dutiona) [reply memory-engine/lib/memory-engine/src/search/hybrid.rs:1013]: `chrono::Duration` is the established majority convention in this codebase (69 uses vs 40 for `TimeDelta` across `src/`) and is a re-exported type alias for `chrono::TimeDelta` — behaviour is identical. The new tests use `Duration` for consistency with that majority and with the already-merged #835 expired-probe tests (which also use `chrono::Duration`). Keeping it as-is; no change. Resolving.

File: memory-engine/lib/memory-engine/src/search/hybrid.rs [hunk 4]
```diff
@@ -877,6 +940,160 @@ mod tests {
...
+        let travel_to = future_valid + chrono::Duration::seconds(1);
```
[the remark this answers is not in the document]
Comment (dutiona) [reply memory-engine/lib/memory-engine/src/search/hybrid.rs:1014]: `chrono::Duration` is the established majority convention in this codebase (69 uses vs 40 for `TimeDelta` across `src/`) and is a re-exported type alias for `chrono::TimeDelta` — behaviour is identical. The new tests use `Duration` for consistency with that majority and with the already-merged #835 expired-probe tests (which also use `chrono::Duration`). Keeping it as-is; no change. Resolving.

**Comment by dutiona** [comment]
**[Adversarial Review] — agy (Gemini 3.1 Pro, pinned diff) + lens subagent (test-quality)**

The lens verified the tests are non-tautological: the temporal facts genuinely reach `assemble_results` (FTS/vector SQL filters only `t_expired`, so future/invalidated facts flow to the post-filter under test), the `rerank_depth` clamp(<limit)/widen(>limit) assertions check the right effective-limit outcomes, and the `rrf_merge` k=0 (rank-0 score = 1.0) + duplicate-id double-count (1/61 + 1/62) arithmetic is correct. Added an overfetch-attrition boundary test (20 post-filter-excluded facts + 1 present survivor). The `chrono::Duration` style threads were resolved (it's the repo-majority convention + a `TimeDelta` alias).
